### DISCUSSION TITLE
Navigating the StoryRiver using KeyboardShortcuts

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#5262] [Ideas]
In #4782 we've discussed broadly the possibility of navigating the story-river using keyboard shortcuts. I'd like to move the discussion here, so that we can gather ideas for 5.1.24 or one of the next releases to tackle this

**Comment by BurningTreeC** [comment]
So ... I'm now at the point where I'd like to tackle this :smile_cat: 

I'll try to create a summary of our previous discussions here

**Reply by pmario** [reply]
Before you start coding please talk to me

**Reply by BurningTreeC** [reply]
Hi @pmario, I'm all ears :smile_cat:

**Reply by BurningTreeC** [reply]
@pmario - what are your ideas, how would you implement this?

**Comment by pmario** [comment]
@BurningTreeC ... Sorry for the delay. 

The last time I did try to tackle the problem, it was top-down from the "tm-navigate" message down to the storyview. It didn't work out too well. 

So now let's reduce it to the keyboard handling and see, where we need to make changes. .. At the moment the interface between the different js-code elements are not well defined and that makes it difficult to change the code. .. That's the reason, why all the existing plugins, that deal with the story "visibility" handling are very hacky. 

I think for the keyboard handling, the only important tiddlers are `$:/StoryList` and `$:/HistoryList` but ... (more to follow)

**Comment by pmario** [comment]
Since TW 5.1.23 we use `-[prefix[$:/StoryList]] -[prefix[$:/HistoryList]]` in the "saver" filters, which will allow us to manage several independent stories. That's a big win, since the last try

**Reply by pmario** [reply]
So the keyboard handling will need to know about the "active" story, that should be addressed with a keyboard shortcut. So every event needs to know, from which "story-source-context" it is activated and which "story-target-context" it needs to modify

**Reply by BurningTreeC** [reply]
Yes, that's what I was thinking, too.

Hi @pmario btw

**Comment by pmario** [comment]
There are 16 messages, that are handled by the navigator widget. 

https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/widgets/navigator.js#L19-L37

I did create a "test-framework" that will allow us to test the TW code interactively, to make sure, that we don't break things. https://wikilabs.github.io/editions/trigger/

**Reply by pmario** [reply]
the most important one atm is `tm-navigate` since it directly activates the whole "animation chain"

**Reply by BurningTreeC** [reply]
Hi @pmario, what are you doing in your test framework? Could you explain it a bit?

**Reply by pmario** [reply]
I basically click the "Next" button, and at the end there should be "no errors" in the output field. This should allow us to check, if we messed something up in the navigator widget

**Reply by BurningTreeC** [reply]
Yeah, but what are the messages the "Next" button is sending? What does it do?

**Reply by pmario** [reply]
From my point of view it sends the `tm-navigate`, with some NEW info about the "target-story-context"

**Reply by pmario** [reply]
`tm-navigate` can be "intercepted" by the linkcatcher-widget, which allows end-users to "re-traget" links to different stories, using wikitext

**Comment by pmario** [comment]
As you can see: 

https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/widgets/navigator.js#L153

the `tm-navigate` message uses `this.addToStory();`, which is defined in navigator.js. `this.addToStory()` looks like this:

https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/widgets/navigator.js#L131

which is defined in story.js

https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/story.js#L37

-------

This dataflow is _not_ as consistent as here, for all of the messages. .. We will need to fix this

**Reply by pmario** [reply]
You know about these functions, since you helped to make the "story" be part of the navigator widget. So we only need to instantiate the code once

**Reply by pmario** [reply]
I think story.js should be to ONLY place that modifies `$:/StoryList???` and `$:/HistoryList???` tiddlers. We need to clean up the API here

**Reply by BurningTreeC** [reply]
right, I know about the functions in the navigator widget

**Comment by pmario** [comment]
As soon as the program hits this function: https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/story.js#L82-L91

The "animation-chane" is triggered with a "store change event"

**Comment by pmario** [comment]
**may be just a reminder**

The startup module `$:/core/modules/startup/story.js`  

"https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/startup/story.js#L21

contains several places, where it deals with the StoryList tiddlers. .. BUT only the default ones are active atm. So we need to think about startup-handling with several stories active at the same tim.

**Reply by BurningTreeC** [reply]
That startup module deals with the main story - do we want to make the main story switchable? Like storing the current story in a config tiddler and using that as the main story dynamically?

**Reply by pmario** [reply]
If 2 or more story rivers are visible at the same time, it may be a "feature request" by users to init both of them properly

**Comment by pmario** [comment]
As written in: https://github.com/Jermolene/TiddlyWiki5/issues/4782#issuecomment-671028109 The HistoryList change triggers the `.handleHistoryChange()` function in the `list-widget.refresh()` 

https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/widgets/list.js#L166

which calls  `storyview.navigateTo()` function.

which is implemented with 5 different `module-tpye: storyview` views atm (3 core and 2 plugins)

The default one is title: $:/core/modules/storyviews/classic.js

The storyviews are essential, because the do the add/remove/navigate animations, which are also needed by the keyboard handlers

**Reply by pmario** [reply]
Storyviews are the only elements, which can be used to tell us about the tiddler positions, which is needed to trigger "prev/next" tiddler actions. 

**At the moment those views don't have an API to give proper feedback about these infos**

**Reply by pmario** [reply]
this line: https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/storyviews/classic.js#L37

Is important too. Since at the moment every function that uses positions to scroll a tiddler into view are extreamly "hacky", since the storyviews have no proper API to give feedback where the **real top of the page** coordinates are

**Comment by pmario** [comment]
In the whole navigation process, there are several widgets involved: 

 - list-widget, which defines **HistoryList, StoryViews,** ViewTemplate and EditTemplate. Changes to HistoryList tirgger the refresh mechanism
 - navigator-widget
 - link-widget
 - linkcatcher-widget
 - keyboard-widget

IMO we must keep an eye on the great context. Otherwise the implementation will be hacky and will make it even more difficult to fix the navigation mess.

**Reply by pmario** [reply]
Let's start with the following widgets

 - link
 - linkcatcher
 - keyboard

--------

Those widgets will need to know about `tv-story-list` and `tv-history-list` ... These variables define the "active-story-context", where the widget is activated

**Reply by pmario** [reply]
I don't know, if the keyboard widget knows about those variables? @BurningTreeC You will ;)

**Reply by pmario** [reply]
It should be possible to use the linkcatcher-widget to switch the "target-story-context", in a defined way. At the moment the parameters are workarounds at best. From my point of view they are hacks.

**Reply by BurningTreeC** [reply]
The keyboard widget is something different from the "global" keyboard-shortcuts mechanism. The global kb-mechanism doesn't know about `tv-story-list` and `tv-history-list` per se - that's why global keyboard shortcuts need a `<$navigator>` widget if they use action-widgets that are handled by the navigator

**Reply by pmario** [reply]
You are right. I always forget about that. ... 

Do "modals" have the same "restrictions"?

**Reply by pmario** [reply]
I think, all the infos, that are needed by the global kb-handling should be stored in the History-tiddlers and may be an new `$:/StoryContext` or `$:/StoryInfo` or similar named tiddler, that tells the global handling, which story to manipulate

**Reply by BurningTreeC** [reply]
I believe recently the modal mechanism was "wrapped" somehow into a navigator widget but I don't recall how

**Reply by pmario** [reply]
Do you think, modals should be usable with global kb-handling? ... Or should this be a "second" step? ... Can it be a "second step"?

**Reply by BurningTreeC** [reply]
We can already trigger modals using global kb shortcuts - what do we want to handle?

**Reply by pmario** [reply]
A modal can contain a "story". So it may be needed to navigate it

**Comment by pmario** [comment]
The linkcatcher is a lightweight (memory / code) widget, that was designed to "trick" the navigator widget. From my point of view it should be used to send the right information to the navigator widget, so it can do its job

**Reply by pmario** [reply]
IMO the same is true for the NEW "keyboard" handlers

**Reply by pmario** [reply]
I do have some prototype code for "link" and "linkcatcher" that sends more info to the navigator which can be used to define the "target-story-context" ... But we need to define the information, that we need first

**Comment by pmario** [comment]
@BurningTreeC .. You may think, that all of this is "overkill". ... I know, but I think it's the big picture, that we have to keep in mind. ... 

Because if the new StoryKeyboard handling is a hack, IMO it will make it impossible to fix the existing navigation API issues.

**Comment by pmario** [comment]
I don't know the keyboard handlers very well. I was happy, that you did all the heavy lifting for the "input / dropdown" navigation. 

I'll have to dig deeper into the code and the new wikitext, which looks super "complex" to me atm.

**Comment by pmario** [comment]
# $:/HistoryList

 - has a text field and
 - `current-tiddler` field atm
   - current-tiddler is written here: Story.prototype.addToHistory = function(navigateTo,navigateFromClientRect) {
   - https://github.com/Jermolene/TiddlyWiki5/blob/5832002feb2b662d65a8d0a38c1c81a390e03317/core/modules/story.js#L100

We probably need more info here

**Reply by pmario** [reply]
## Proposal

I suggest, that there should be a possibility, that plugins and new keyboard handler functions should be able to add arbitrary fields here, that will trigger a "history-refresh" in the list-widget.

**Reply by pmario** [reply]
eg: with a "selected-tiddler" or "active-tiddler" it will be easy to implement some extra highlighting in the TOC, or the "story river".

**Comment by pmario** [comment]
# $:/StoryList

 - has the `list` field, which can probably be used to define "next / previews" function

**Reply by pmario** [reply]
- The $:/StoryList ~~widget~~ is highly connected with the "navigator-widget". 
 - The $:/HistoryList and StoryViews are highly connected with the "list-widget"

**Reply by BurningTreeC** [reply]
What do you mean with "StoryList widget"?

**Reply by pmario** [reply]
uups ... fixed the text above. It should be $:/StoryList tiddler

**Comment by pmario** [comment]
# Active Story

 - IMO the global keyboard handling needs some info, which story is the "active" one, that should be manipulated

**Reply by BurningTreeC** [reply]
Yes. I think that must be done with a workaround through a state/config tiddler that stores the "active/current" story

**Reply by pmario** [reply]
May be there need to be 2 or more of them. 

 - Default: As it is now
   - Default - Is used if there is no config-tiddler, for backwards compatibility
 - Config - Is used for startup and if no state tiddler exists
 - States are not saved, so config is used for startup

We may need a configuration somewhere, that allows us to init 2 or more stories at startup. 

Let's suggest some names!?

**Reply by BurningTreeC** [reply]
$:/config/StoryList

**Reply by BurningTreeC** [reply]
... and $:/config/HistoryList respectively

**Reply by pmario** [reply]
Yea, but there may be a unique "qualified" name attached eg: `$:/config/StoryList-189435334534` 

May be:

```
title: $:/config/StoryList-189435334534
text: #NavigatorWidget:NavigatorWidget ActionNavigateWidget ListWidget [[Draft of 'New Tiddler']] $:/StoryList
```

using the same format as a TW permalink: https://tiddlywiki.com/#NavigatorWidget:NavigatorWidget%20ActionNavigateWidget%20ListWidget%20%5B%5BDraft%20of%20'New%20Tiddler'%5D%5D%20%24%3A%2FStoryList  but without the URL escape chars

**Reply by pmario** [reply]
> 
> 
> ... and $:/config/HistoryList respectively

I'm not sure, if we need s config for HistoryList. What do you want to store there?

**Reply by BurningTreeC** [reply]
>I'm not sure, if we need s config for HistoryList. What do you want to store there?

The title of the HistoryList corresponding with the current storyList

**Reply by pmario** [reply]
May be a data tiddler for the state:

```
title: $:/state/active/Story

history: $:/HistoryList-1
story: $:/StoryList-1
```

**Reply by BurningTreeC** [reply]
Hmm. That is something to decide another day. But we have common sense that both need to be stored somewhere

**Comment by BurningTreeC** [comment]
@pmario - I DON'T think this is an overkill but I'm missing an overview about the changes you have in mind

It would be helpful if you could provide some starting points like "the current story should be stored in a config or state tiddler", "the currently active/selected tiddler should be stored in the $:/HistoryList tiddler". "The problem that arises from that is that ..."

I currently find it a bit confusing knowing that you have a clearer picture in your mind than you are depicting here

**Reply by pmario** [reply]
>... I currently find it a bit confusing knowing that you have a clearer picture in your mind than you are depicting here

Yes. I think every list is a story. ... But the current UI doesn't use it that way, because we don't have proper "multi story" handling in the core. 

Once we have that, I'm sure our UI wikitext can be much simpler and much more powerful

**Comment by pmario** [comment]
# Navigate multiple Stories

The navigator-widget has the following parameters: 

Attribute | Description
-- | --
story | Name of the tiddler containing the story list to be manipulated
history | Name of the tiddler containing the history list to be manipulated
openLinkFromInsideRiver | Determines the location for opening new tiddlers from links within the story river: at the top or bottom of the story river, or above or below the current tiddler
openLinkFromOutsideRiver | Determines the location for opening new tiddlers from linkes outside the story river: at the top or bottom of the story river

**story** and **history** are defined whith the widget call in wikitext. 

**The other params** should be configurable within the control panel or a different UI

@BurningTreeC ... "The other params" may also be important for the global kb-handling

**Reply by pmario** [reply]
IMO 

Attribute | Description
-- | --
relinkOnRename |  Determines if tags in other tiddlers are renamed if the tiddler title changes. yes or no. default: no

should be **global ONLY**

**Comment by BurningTreeC** [comment]
please @pmario create a SUMMARY of your ideas

**Reply by pmario** [reply]
I'll try. But I think https://github.com/Jermolene/TiddlyWiki5/discussions/5262#discussioncomment-274994 is the most important one atm, that may be written into an actionable "proposal issue"

**Comment by Jermolene** [comment]
Thanks @pmario it would indeed be useful to have a clear summary of the proposal, and perhaps also to enumerate the requirements the proposal is trying to meet.

I'm concerned about the attention to multiple story rivers. That's definitely a feature that we should add to the core, but I'm not sure that we need to be so concerned with it here. This is about maintaining a visible focus which can be moved via the keyboard or mouse, and keyboard actions that apply to the focused element. We definitely need to be able to track the focus moving to elements outside the main story river (eg elements in the menu bar or side bar), but I don't think that those elements would necessarily be tiddlers (or "tiddler frames" to be clearer), nor would they necessarily be in a river.

So, there wouldn't be a keyboard shortcut for "edit tiddler", there'd be a keyboard shortcut for "edit current thing" which would adapt according to the element that has the focus.

We will need keyboard shortcuts that affect the current story river: the `<$keyboard>` widget would wrap the story river, and trap the required shortcuts, just as we do for tiddlers.

For me, the underpinning for all of this is the browser implementation of focus. As we've discussed before, I think we should enhance the refresh cycle so that it preserves the focussed element, and tracks the focussed element via state tiddlers.

When we introduce multiple story rivers we will certainly need state tiddlers to keep track of things like which story river should be used to open new tiddlers, but I'd be hoping we wouldn't need to worry about that yet.

**Reply by saqimtiaz** [reply]
@Jermolene agreed. I don't see why we can't introduce keyboard navigation of the story now and deal with multiple stories when we add them. The sticky bit is as you mention, keeping track of the focus element and restoring focus after refresh. 

I actually think @BurningTreeC's first implementation wasn't far off from what we need.

**Reply by BurningTreeC** [reply]
Thanks for your feedback here, @Jermolene 

>For me, the underpinning for all of this is the browser implementation of focus. As we've discussed before, I think we should enhance the refresh cycle so that it preserves the focussed element, and tracks the focussed element via state tiddlers.

I know you mentioned this before and I now have a slightly clearer picture of what it should do and why it's useful. But doesn't that mean we'd have to make all affected elements focusable through e.g. the `tabindex` attribute?

I currently also don't know where to look for the refresh-cycling handling. I'd appreciate a hint in which file to look for it

<hr>

About multi-story awareness I believe big parts of it can be done in wikitext and only where we reach its limits we should make adjustments in js.

<hr>

>We will need keyboard shortcuts that affect the current story river: the <$keyboard> widget would wrap the story river, and trap the required shortcuts, just as we do for tiddlers.

How can the keyboard widget trap shortcuts that aren't issued within an input or textarea? You certainly have an idea how to use it for this case - but I do not

**Reply by Jermolene** [reply]
Hi @BurningTreeC 

> I know you mentioned this before and I now have a slightly clearer picture of what it should do and why it's useful. But doesn't that mean we'd have to make all affected elements focusable through e.g. the `tabindex` attribute?

I think that's right, yes.

> I currently also don't know where to look for the refresh-cycling handling. I'd appreciate a hint in which file to look for it

The refreshing of the main story river happens in https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/render.js

However, one of the challenges is that the refresh handling is replicated in a few places in the code base (for example, external windows are handled in https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/windows.js). So, a prerequisite might be to refactor things to give us a generic class/object for rendering and handling subsequent refreshes that we can reuse across the core.

> About multi-story awareness I believe big parts of it can be done in wikitext and only where we reach its limits we should make adjustments in js.

That's my thinking too.

> > We will need keyboard shortcuts that affect the current story river: the <$keyboard> widget would wrap the story river, and trap the required shortcuts, just as we do for tiddlers.
> 
> How can the keyboard widget trap shortcuts that aren't issued within an input or textarea? You certainly have an idea how to use it for this case - but I do not

The DOM handles that. If the tiddler frame is focussed then it will get keyboard events. Any events not handled will get offered to ancestor nodes, one of which would be the keyboard widget around the story river.

---

We'll need to split this up something like this:

1. Refactor refresh handling into one place
2. Improve refresh handling to preserve the focus
3. Add keyboard shortcuts for the story river and for individual tiddlers

**Reply by BurningTreeC** [reply]
Thanks for your feedback @saqimtiaz ,


>@Jermolene agreed. I don't see why we can't introduce keyboard navigation of the story now and deal with multiple stories when we add them. The sticky bit is as you mention, keeping track of the focus element and restoring focus after refresh.

I think @pmario is already thinking about users that add their own stories, but I agree that we can think about that later.
Would you mind elaborating a bit (for my understanding) what keeping track of the focus element and restoring its focus after refresh means in this context? In which cases is it needed? How does the refresh cycle affect the currently focused element?

**Reply by saqimtiaz** [reply]
> However, one of the challenges is that the refresh handling is replicated in a few places in the code base (for example, external windows are handled in https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/windows.js). So, a prerequisite might be to refactor things to give us a generic class/object for rendering and handling subsequent refreshes that we can reuse across the core.

Modals might be another one that we need to look at for this.

@BurningTreeC Regarding refresh and understanding how it works, it may help to pay attention to how refresh() is called for the pageWidget node (which is a transclusion of the page template): https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/render.js#L75

Note how each widget calls refresh for its children: https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/widget.js#L447

I am going to be tied up for another two weeks or so, but if you @BurningTreeC  like I can work with you on this after that, much as we did for the keyboard-driven-inputs.

**Reply by saqimtiaz** [reply]
> Would you mind elaborating a bit (for my understanding) what keeping track of the focus element and restoring its focus after refresh means in this context? In which cases is it needed? How does the refresh cycle affect the currently focused element?

@BurningTreeC  imagine a story river with the tiddlers:

- A
- B
- C

Imagine that B is the tiddler that has focus, and now you click a button in that tiddler that makes tiddler B get refreshed in the story. This means the DOM node for the tiddler content will be destroyed and re-created to refresh it. After this refresh, tiddler B will no longer have focus, the focus was lost when the DOM node was destroyed before being re-created. 

So as a simple example we need to remember which tiddler had focus, and then restore that after refresh. Ideally we want to take it a step further and restore focus to the exact element which had focus, which might be a button or an input field in a tiddler.

As background note how the transclude widget calls refreshSelf() when the tiddler being transcluded has changed:
https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/transclude.js#L107
 
And then note how refreshSelf() destroys the widgets DOM nodes and then calls render() to recreate them:
https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/widget.js#L455

**Reply by Jermolene** [reply]
> I think @pmario is already thinking about users that add their own stories, but I agree that we can think about that later.
> Would you mind elaborating a bit (for my understanding) what keeping track of the focus element and restoring its focus after refresh means in this context? In which cases is it needed? How does the refresh cycle affect the currently focused element?

The "refresh cycle" is the process that updates a rendering when there are changes in the tiddler store. It involves recursively calling the `refresh()` method of each widget in the tree, giving them a chance to respond to changes in the tiddler store. In doing so, they will create and delete DOM nodes as required. The problem comes if one of those DOM nodes had the focus; when it is deleted the browser will follow it's own rules to assign the focus elsewhere, but it won't automatically re-assign the focus to any replacement element.

So the gist of what we need to do is to update the refresh cycle so that first we somehow record which widget has the focus, and then after the refresh cycle we restore it. As I say, the complication is that some or all of the widget and DOM trees might be destroyed and recreated in the course of the refresh. That means that we can't store state data in the DOM tree or the widget tree.

**Reply by BurningTreeC** [reply]
Hi @Jermolene ,

Thanks, you made it all clear!

**Reply by BurningTreeC** [reply]
@Jermolene @saqimtiaz ,

Thank you both for your clear explanations!

**Reply by BurningTreeC** [reply]
@saqimtiaz 

>I am going to be tied up for another two weeks or so, but if you @BurningTreeC like I can work with you on this after that, much as we did for the keyboard-driven-inputs.

I'd like to do so very much. In the meantime I'm looking into the code to understand the inner workings and to (maybe) fiddle around with it

**Comment by pmario** [comment]
I think it's much easier than you think. The only thing the global story handling needs to konw is, which is the "target story context" that should be modified. Then it needs to modify the $:/History-xxx tiddler and all the magic should happen automatically.

The only difficulties I see, is to combine mouse and keyboard actions. Eg. The user may want to scroll down with the mouse wheel. Then eg: use PageUp key to scroll ti the top if the the "active" tiddler if it is a long tiddler. If the tiddler is short, PageUp should scroll to the top of the previous tiddler. ... and so on. 

If the mouse wheel is scrolled down or up in the meantime, the focus of the tiddler needs to change. 

I think ALL if this can be done already with no problems, for the hardcoded single story we have. BUT it gets extreamly hacky as soon as we start to introduce several stories. 

I'll try to create a prototype, to show what I mean.

**Reply by Jermolene** [reply]
> I think it's much easier than you think. The only thing the global story handling needs to konw is, which is the "target story context" that should be modified. Then it needs to modify the $:/History-xxx tiddler and all the magic should happen automatically.

I'm having trouble following your meaning here. What is a target story context? What should modify it? What is the "it" that modifies the history list tiddler? What magic happens?

> The only difficulties I see, is to combine mouse and keyboard actions. Eg. The user may want to scroll down with the mouse wheel. Then eg: use PageUp key to scroll ti the top if the the "active" tiddler if it is a long tiddler. If the tiddler is short, PageUp should scroll to the top of the previous tiddler. ... and so on.

> If the mouse wheel is scrolled down or up in the meantime, the focus of the tiddler needs to change.

We shouldn't re-invent how web apps work. I don't think that scrolling shouldn't change the keyboard focus.

> I think ALL if this can be done already with no problems, for the hardcoded single story we have. BUT it gets extreamly hacky as soon as we start to introduce several stories.

> I'll try to create a prototype, to show what I mean.

I suspect that it might be a better use of time to try to document your thinking better so that others can follow along. The implementation of this is likely to be complex and hard to follow. Given that at this point I don't fully understand the goals of your design I would find reviewing the code nearly impossible.

**Reply by pmario** [reply]
At the moment we do have several parameters that can be used to configure the behaviour of the story: https://github.com/Jermolene/TiddlyWiki5/discussions/5262#discussioncomment-275049

 - At the moment these parameters belong to, what I call a "Story Context": **story, history, openLinkFromInsideRiver and openLinkFromOutsideRiver**
 - Those elements are defined using the `<$navigator` widget.
 - navigator.js internally defines: `tv-story-list` and `tv-history-list` .. So they are also part of the "Story Context"

**Reply by pmario** [reply]
@Jermolene 

>I'm having trouble following your meaning here. .... What should modify it?

The "story handling" should modify the $:/HistoryList-xxxx tiddler. Story handling can be done by mouse-wheel, mouse-clicks, global keyboard shortcuts and local shortcuts (inside keyboard-widget body)

> ... What is the "it" that modifies the history list tiddler? 

"story handling"

> ... What magic happens?

I did some debugging some time ago and I was sure, that the $:/HistoryList "change" event will trigger the story "scroll into view" animation. ... which I called "magic". ... At the moment that's not the case. So no magic happens. ... I'll need to do a bit more code investigation here

**Comment by BurningTreeC** [comment]
I have a stupid question regarding the preservation of the focused element:

I guess we want to do this using `document.activeElement` ... Now, to preserve focus over a refresh cycle, should a) the information about the active element be stored in a state tiddler or b) can we store it in a global variable? And if a) - don't we need a unique identifier for each widget?

**Reply by BurningTreeC** [reply]
I think the `generateTransclusionFootprint` method which I tried to introduce in  #4965 could work for this. You @Jermolene did mention that it can be used for focus preservation/restoring.

**Reply by Jermolene** [reply]
Hi @BurningTreeC I think we shouldn't attempt to track the currently focussed element in a state tiddler. Instead, before the refresh cycle we should identify the element with focus and store a reference to it in a JS variable, and then once the refresh is done we identify the element that should have focus (which may be a different element, but will have the same "render tree footprint", which might be a better name for "generateTransclusionFootprint").

**Reply by BurningTreeC** [reply]
>Hi @BurningTreeC I think we shouldn't attempt to track the currently focussed element in a state tiddler. Instead, before the refresh cycle we should identify the element with focus and store a reference to it in a JS variable, and then once the refresh is done we identify the element that should have focus (which may be a different element, but will have the same "render tree footprint", which might be a better name for "generateTransclusionFootprint").

Thanks. Yes, render tree footprint is a better name for it. I have played with the code in the meantime and have a better understanding now. I'm already able to restore focus of a button widget for example but my method would force us to edit all widgets with focusable domNodes. Do you think that's the way to go?

- assigning a render-tree-footprint to each widget and storing the footprint besides the widget in a global array of objects
- before the refresh-cycle storing the activeElement globally
- when a widget renders detecting if its render-tree-footprint corresponds with the activeElement (I'm already able to do so)

**Reply by Jermolene** [reply]
> Thanks. Yes, render tree footprint is a better name for it. I have played with the code in the meantime and have a better understanding now. I'm already able to restore focus of a button widget for example but my method would force us to edit all widgets with focusable domNodes. Do you think that's the way to go?

That certainly seems like one possibility, but it might be interesting to try to do this without requiring widgets to be modified. Here's one approach that might work. To save the reference to the focussed element:

1. Find the DOM node with the focus
2. Locate the widget the owns that DOM node (widgets have the `domNodes` array that lists the nodes it has created, but one also needs to worry about any children that they might have)
3. Store the reference in two parts: a render tree footprint identifying the widget, and then some numbers identifying the DOM node within the widget

> * assigning a render-tree-footprint to each widget and storing the footprint besides the widget in a global array of objects

I don't think we need to pre-compute the render tree footprint, it's really just the state data for walking the render tree. We'd map the render tree footprint to the render tree by walking the render tree.

> * before the refresh-cycle storing the activeElement globally

As above.

> * when a widget renders detecting if its render-tree-footprint corresponds with the activeElement (I'm already able to do so)

We can instead have the refresh mechanism use normal DOM methods to set the focus at the end of the cycle.

**Reply by BurningTreeC** [reply]
Thank you @Jermolene ,

> > Thanks. Yes, render tree footprint is a better name for it. I have played with the code in the meantime and have a better understanding now. I'm already able to restore focus of a button widget for example but my method would force us to edit all widgets with focusable domNodes. Do you think that's the way to go?
> 
> That certainly seems like one possibility, but it might be interesting to try to do this without requiring widgets to be modified. Here's one approach that might work. To save the reference to the focussed element:
> 
> 1. Find the DOM node with the focus
> 2. Locate the widget the owns that DOM node (widgets have the `domNodes` array that lists the nodes it has created, but one also needs to worry about any children that they might have)
> 3. Store the reference in two parts: a render tree footprint identifying the widget, and then some numbers identifying the DOM node within the widget
> 

ah, I understand. Wouldn't that require us to walk DOWN the widget tree recursively?

> > * assigning a render-tree-footprint to each widget and storing the footprint besides the widget in a global array of objects
> 
> I don't think we need to pre-compute the render tree footprint, it's really just the state data for walking the render tree. We'd map the render tree footprint to the render tree by walking the render tree.
> 
> > * before the refresh-cycle storing the activeElement globally
> 
> As above.
> 
> > * when a widget renders detecting if its render-tree-footprint corresponds with the activeElement (I'm already able to do so)
> 
> We can instead have the refresh mechanism use normal DOM methods to set the focus at the end of the cycle.
> 

Yes I can see that that's possible

**Reply by Jermolene** [reply]
> ah, I understand. Wouldn't that require us to walk DOWN the widget tree recursively?

Yes, although perhaps it's arbitrary whether it's up or down :smile:

I've some old code that does almost what is needed:

```js
// Find the child widget containing the event target
var findWidgetOwningDomNode = function(widget) {
	for(var domNodeIndex=0; domNodeIndex<widget.domNodes.length; domNodeIndex++) {
		if(widget.domNodes[domNodeIndex] === domNode) {
			return widget;
		}
	}
	for(var childIndex=0; childIndex<widget.children.length; childIndex++) {
		var result = findWidgetOwningDomNode(widget.children[childIndex]);
		if(result) {
			return result;
		}
	}
	return null;
};
var widget = findWidgetOwningDomNode($tw.rootWidget);
```

**Reply by BurningTreeC** [reply]
Thank you for the code example, it will help me a lot walking the widget tree DOWN or UP or whatever :grin:

This is a very valuable discussion thread for me, thank you @Jermolene !

**Reply by BurningTreeC** [reply]
P.S. I would probably have written similar code after hours of thinking ;)

**Reply by BurningTreeC** [reply]
Hi @Jermolene , I now have a working prototype that preserves focus over refresh cycles like you proposed. My next question would be regarding how the `<$keyboard>` widget should know about the tiddler that has focus or focus within.

**Reply by Jermolene** [reply]
> Hi @Jermolene , I now have a working prototype that preserves focus over refresh cycles like you proposed. My next question would be regarding how the `<$keyboard>` widget should know about the tiddler that has focus or focus within.

I don't think it needs to. The keyboard handlers for the current tiddler would be wrapped around the tiddler frame, as we do in $:/core/ui/EditTemplate, and the keyboard handlers for the current story would be wrapped around the story river.

**Reply by BurningTreeC** [reply]
> > Hi @Jermolene , I now have a working prototype that preserves focus over refresh cycles like you proposed. My next question would be regarding how the `<$keyboard>` widget should know about the tiddler that has focus or focus within.
> 
> I don't think it needs to. The keyboard handlers for the current tiddler would be wrapped around the tiddler frame, as we do in $:/core/ui/EditTemplate, and the keyboard handlers for the current story would be wrapped around the story river.
> 

Oh, right! Thanks!

Another thing that could be problematic with this implementation: If I click the new-tiddler button, focus is on the button and will be restored to it. The new tiddler that was then created will not have focus in the title field right after that. There are more subtleties this implementation brings with it

**Reply by Jermolene** [reply]
> Another thing that could be problematic with this implementation: If I click the new-tiddler button, focus is on the button and will be restored to it. The new tiddler that was then created will not have focus in the title field right after that. There are more subtleties this implementation brings with it

Yes, it's an interesting case! At the moment, the way we do this relies on the fact that if a new element is added to the DOM that has the "autofocus" attribute set then browsers automatically move the focus to it.

One could imagine giving widgets a method for them to call during render or refresh to indicate that they want to be given the focus at the end of the current render/refresh cycle. We'd need to have some mechanism for arbitrating between competing claims for the focus.

**Reply by BurningTreeC** [reply]
>One could imagine giving widgets a method for them to call during render or refresh to indicate that they want to be given the focus at the end of the current render/refresh cycle. We'd need to have some mechanism for arbitrating between competing claims for the focus.

That brings me back to the point where I want to edit all affected widgets and detect within their render method if they should focus their domNode again

That would then force me to generate the renderTreeFootprint within each widget but as you mentioned - that shouldn't be the way to go

**Comment by BurningTreeC** [comment]
A little summary of my current state:

- before the refresh cycle begins I get the widget that owns the currently focused domNode
- then I generate that widgets' `renderTreeFootprint` that's an array of indexes
- I also store the index of the domNode in the widgets' `domNodes` array
- after the refresh cycle I walk the widget tree from the `rootWidget` using the stored indexes to identify the widget that owns the domNode that should be focused again
- but in the meantime the widget can be totally destroyed and also its parent widgets, so that walking the tree results in an undefined widget
  - there seems to be the need of using wikitext to focus elements in certain cases
  - for example when editing a tiddler and saving it, the focus will not automatically be restored to an element within the tiddler or the tiddler-frame itself
  - I haven't yet tried if it's possible to set focus to the then-to-be-created tiddler in view mode through an action on the save-tiddler button but I doubt that it works

What I'm thinking of now is how the mechanism should know which tiddler or element to focus again after the refresh if at the renderTreeFootprint there's no widget anymore

**Reply by Jermolene** [reply]
Hi @BurningTreeC that's great, it sounds like it might be worth pulling together a draft PR at this point so that we can review?

**Reply by BurningTreeC** [reply]
Hi @Jermolene , I'll create a draft PR this weekend.

**Reply by BurningTreeC** [reply]
@Jermolene - I've put together a Draft PR in #5414
